docs/workflow/testing/libraries/testing.md

```diff
@@ -5,80 +5,88 @@
 These example commands automate the test run and all pre-requisite build steps in a single command from a clean enlistment.
 
 - Run all tests - Builds clr in release, libs+tests in debug:
+
 ```
 build.cmd/sh -subset clr+libs+libs.tests -test -rc Release
 ```
 
 - Run all tests - Builds Mono in release, libs+tests in debug:
+
 ```
 build.cmd/sh -subset mono+libs+libs.tests -test -rc Release
 ```
 
 - Run all tests - Build Mono and libs for x86 architecture in debug (choosing debug for runtime will run very slowly):
+
 ```
 build.cmd/sh -subset mono+libs+libs.tests -test -arch x86
 ```
 
 ## Partial Build and Test Runs
 
-Doing full build and test runs takes a long time and is very inefficient if you need to iterate on a change.
-For greater control and efficiency individual parts of the build + testing workflow can be run in isolation.
-See the [Building instructions](../../building/libraries/README.md) for more info on build options.
+Doing full build and test runs takes a long time and is very inefficient if you need to iterate on a change. For greater control and efficiency individual parts of the build + testing workflow can be run in isolation. See the [Building instructions](../../building/libraries/README.md) for more info on build options.
 
 ### Test Run Pre-requisites
-Before any tests can run we need a complete build to run them on. This requires building (1) a runtime, and
-(2) all the libraries. Examples:
+
+Before any tests can run we need a complete build to run them on. This requires building (1) a runtime, and (2) all the libraries. Examples:
 
 - Build release clr + debug libraries
+
 ```
 build.cmd/sh -subset clr+libs -rc Release
 ```
 
 - Build release mono + debug libraries
+
 ```
 build.cmd/sh -subset mono+libs -rc Release
 ```
 
-Building the `libs` subset or any of individual library projects automatically copies product binaries into the testhost folder
-in the bin directory. This is where the tests will load the binaries from during the run. However System.Private.CorLib is an
-exception - the build does not automatically copy it to the testhost folder. If you [rebuild System.Private.CoreLib](https://github.com/dotnet/runtime/blob/main/docs/workflow/building/libraries/README.md#iterating-on-systemprivatecorelib-changes) you must also build the `libs.pretest` subset to ensure S.P.C is copied before running tests.
+Building the `libs` subset or any of individual library projects automatically copies product binaries into the testhost folder in the bin directory. This is where the tests will load the binaries from during the run. However System.Private.CorLib is an exception - the build does not automatically copy it to the testhost folder. If you [rebuild System.Private.CoreLib](https://github.com/dotnet/runtime/blob/main/docs/workflow/building/libraries/README.md#iterating-on-systemprivatecorelib-changes) you must also build the `libs.pretest` subset to ensure S.P.C is copied before running tests.
```

Comment:
I think there are indeed two facets to this discussion - (a) using Crossgen2-compiled System.Private.CoreLib in library testing and (b) compiling the framework and / or the library tests themselves using Crossgen2. For (a), AFAIK that's always been the case, to my knowledge no testing involves the IL version of SPC. For (b), that's what Ivan is now creating foundations for unless I'm mistaken as at least weekly library lab testing involving Crossgen2 would be very beneficial for improving our test coverage; @davidwrighton also believes Crossgen2 library testing could be used for better regression monitoring of some of our features like the cross-module inlining. Having said that, I'm not sufficiently familiar with library testing to know off the top of my head whether the comment under discussion is actually accurate or not.

Reply:
Thanks Viktor for clarifying the usage of IL SPC. For the rest of your comment, so am I right to understand your point is about "Native-AOT compiled self-contained Crossgen2 app" vs. "raw framework-dependent Crossgen2 app produced by CoreCLR build"? I certainly agree with that, we should do our best to test the actual customer-facing product, not some intermediate artifact of our build system (not mentioning the fact that in the past it also identified bugs in NativeAOT). I apologize, I didn't get that from your original PR comment at the top of this thread, in fact it seems you're saying the opposite in

> we want to exercise Crossgen2 testing in self-contained deployment only in the dev innerloop

I apologize if I misread or misunderstood one of your responses, in such case please have patience with me being slow on the uptake and clarify.

Reply:
> I don't think we want to do framework dependent crossgen2 testing but self-contained deployment testing.

crossgening the whole framework takes a while. Would self-contained deployment mean that the whole framework needs to be crossgened for each libraries test? It does not look very efficient.

Reply:
In CoreCLR, crossgenning the framework takes something like 2~3 minutes. It should certainly be done only once, not in Helix for each test separately, that would be hugely wasteful.

Reply:
> crossgening the whole framework takes a while. Would self-contained deployment mean that the whole framework needs to be crossgened for each libraries test? It does not look very efficient.

No. It would mean that we could crossgen the runtime pack that we use for self-contained deployment once, instead of the testhost folder that we use for framework-dependent deployment. It boils down to which app-mode we want to add crossgen validation to. SFF, SCD or both?